src/fu-dbus-daemon.c

```diff
@@ -1685,25 +1685,56 @@ fu_dbus_daemon_method_emulation_save(FuDbusDaemon *self,
 				  g_steal_pointer(&helper));
 }
 
+static void
+fu_dbus_daemon_authorize_modify_device_cb(GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	FuEngine *engine;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(FuMainAuthHelper) helper = (FuMainAuthHelper *)user_data;
+
+	engine = fu_daemon_get_engine(FU_DAEMON(helper->self));
+	if (!fu_engine_modify_device(engine,
+				     helper->device_id,
+				     helper->key,
+				     helper->value,
+				     &error)) {
+		fu_dbus_daemon_method_invocation_return_gerror(helper->invocation, error);
+		return;
+	}
+	g_dbus_method_invocation_return_value(helper->invocation, NULL);
+}
+
 static void
 fu_dbus_daemon_method_modify_device(FuDbusDaemon *self,
 				    GVariant *parameters,
 				    FuEngineRequest *request,
 				    GDBusMethodInvocation *invocation)
 {
-	FuEngine *engine = fu_daemon_get_engine(FU_DAEMON(self));
 	const gchar *device_id;
 	const gchar *key = NULL;
 	const gchar *value = NULL;
-	g_autoptr(GError) error = NULL;
+	g_autoptr(FuMainAuthHelper) helper = NULL;
 
-	/* check the id exists */
 	g_variant_get(parameters, "(&s&s&s)", &device_id, &key, &value);
-	if (!fu_engine_modify_device(engine, device_id, key, value, &error)) {
-		fu_dbus_daemon_method_invocation_return_gerror(invocation, error);
-		return;
-	}
-	g_dbus_method_invocation_return_value(invocation, NULL);
+	helper = g_new0(FuMainAuthHelper, 1);
+	helper->self = self;
+	helper->request = g_object_ref(request);
+	helper->invocation = g_object_ref(invocation);
+	helper->device_id = g_strdup(device_id);
+	helper->key = g_strdup(key);
+	helper->value = g_strdup(value);
+	if (g_strcmp0(key, "Flags") == 0 && g_strstr_len(value, -1, "emulation-tag") != NULL) {
```

Comment:
I'm confused why g_strstr_len? I think you can only modify one flag at a time, no? Os if this to catch `emulation-tag` and `~emulation-tag`? In which case two g_strcmps is probably safer.

Reply:
That's exactly why - to catch `emulation-tag` and `~emulation-tag`.

Reply:
Maybe splitting out a `gboolean fu_dbus_daemon_method_modify_device_flag_needs_auth(key, value)` might be a good idea?

Reply:
added